ros_ws/src/autonomous/wallfollowing2/script/wallfollowing.py

```diff
@@ -0,0 +1,153 @@
+    message = drive_param()
+    message.angle = angle
+    message.velocity = velocity
+    drive_parameters_publisher.publish(message)
+
+
+def laser_callback(message):
+    global laser_scan
+    laser_scan = message
+
+
+def get_scan_as_cartesian():
+    if laser_scan is None:
+        raise Exception("No scan has been received yet.")
+
+    ranges = np.array(laser_scan.ranges)
+    angles = np.linspace(
+        laser_scan.angle_min,
+        laser_scan.angle_max,
+        ranges.shape[0])
+
+    points = np.zeros((ranges.shape[0], 2))
+    points[:, 0] = -np.sin(angles) * ranges
+    points[:, 1] = np.cos(angles) * ranges
+
+    return points
+
+
+def find_left_right_border(points, margin_relative=0.1):
+    margin = int(points.shape[0] * margin_relative)
+
+    relative = points[margin + 1:-margin, :] - points[margin:-margin - 1, :]
+    distances = np.linalg.norm(relative, axis=1)
+
+    return margin + np.argmax(distances) + 1
+
+
+def follow_walls(left_circle, right_circle):
+    predicted_car_position = Point(0, 1.5)
+    left_point = left_circle.get_closest_point(predicted_car_position)
+    right_point = right_circle.get_closest_point(predicted_car_position)
+
+    target_position = Point(
+        (left_point.x + right_point.x) / 2,
+        (left_point.y + right_point.y) / 2)
+    error = target_position.x - predicted_car_position.x
+    if math.isnan(error) or math.isinf(error):
+        error = 0
+
+    steering_angle = pid.update_and_get_correction(
+        error, 1.0 / UPDATE_FREQUENCY)
+
+    radius = min(left_circle.radius, right_circle.radius)
+    speed_limit_radius = (radius - RADIUS_SMALL) / (RADIUS_BIG - RADIUS_SMALL)
+    speed_limit_error = 1 + ERROR_DEAD_ZONE - abs(error) * ERROR_SPEED_DECREASE
+    relative_speed = max(0.0, min(1.0, speed_limit_error, speed_limit_radius))
+    speed = SLOW + (FAST - SLOW) * relative_speed
+
+    drive(steering_angle * (1.0 - relative_speed), speed)
+
+    show_line_in_rviz(2, [left_point, right_point],
+                      color=ColorRGBA(1, 1, 1, 0.3), line_width=0.005)
+    show_line_in_rviz(3, [Point(0, 0), predicted_car_position],
+                      color=ColorRGBA(1, 1, 1, 0.3), line_width=0.005)
+    show_line_in_rviz(4, [predicted_car_position,
+                          target_position], color=ColorRGBA(1, 0.4, 0, 1))
+
+
+def handle_scan():
+    if laser_scan is None:
+        return
+
+    points = get_scan_as_cartesian()
+    split = find_left_right_border(points)
+
+    right_wall = points[:split, :]
+    left_wall = points[split:, :]
+
+    left_circle = circle.fit(left_wall)
+    right_circle = circle.fit(right_wall)
+
+    follow_walls(left_circle, right_circle)
+
+    show_circle_in_rviz(left_circle, left_wall, 0)
+    show_circle_in_rviz(right_circle, right_wall, 1)
+
+
+laser_scan = None
+
+rospy.Subscriber("/scan", LaserScan, laser_callback)
+drive_parameters_publisher = rospy.Publisher(
+    "/input/drive_param/autonomous", drive_param, queue_size=1)
```

Comment:
You could move this string to the to of the file and call the variable TOPIC_DRIVE_PARAMETERS (just like in wall_following.h in wallfollowing1). However that's not really a must.